### ISSUE TITLE
Upcoming Chromium changes (declarativeNetRequest)

### ISSUE THREAD

**Issue by Parasrah** [EFForg/https-everywhere#17268]
Type: code issue

[//]: # (Include any other relevant information below. Thank you again for)
[//]: # (helping to improve HTTPS Everywhere.)

Hi, I've recently read a string of reports about the chromium project moving to a new API called `declarativeNetRequest`, and removing the capabilities of the `webRequest` API to block or redirect requests. From what I understand `declarativeNetRequest` will still allow blocking and redirect rules, but it will not be as flexible as `webRequest` in that you will essentially be providing a fairly simple rule-set and will no longer be able to react to a request programatically. They claim this is to increase the privacy for their users and to make redirects/blocking more efficient, although it may also conveniently (for them) break the functionality of many ad-blockers (apparently according to their authors in the reports I've read, I could not find a source for this).

I was wondering how this will affect HTTPS Everywhere (I know you make extensive use of the `webRequest` API), and if the extension will still be able to function in Chromium. I've also seen speculation that the number of rule sets permitted with `declarativeNetRequest` would be limited, which would obviously be a problem for this extension if the limit was too low.

[Draft of Manifest V3 for chrome extensions](https://docs.google.com/document/d/1nPu6Wy4LWR66EFLeYInl3NzzhHzc-qnk4w4PX-0XMw8/edit#heading=h.5ottncv8stov)

[PDF (for those who would like to avoid google docs)](https://github.com/EFForg/https-everywhere/files/2787975/Manifest.V3.pdf)

[declarativeNetRequest API](https://developer.chrome.com/extensions/declarativeNetRequest)

Thanks for all the work the team has put in to enhance the security of their users, hopefully there is a way around these changes

**Comment by Parasrah** [comment]
Update: I see where the speculation on the maximum number of rules came from, the [api docs](https://developer.chrome.com/extensions/declarativeNetRequest#property-MAX_NUMBER_OF_RULES) mention it to be 30,000

**Comment by zoracon** [comment]
@Parasrah Thank you for bringing this to attention. We had some discussion yesterday about how this impacts the extension as the Manifest proposal currently stands. The main concerns are around:

- Rules handling would be statically handled and not programmatically as mentioned above
- Transference from background scripts to service workers could potentially allow sites to have more control over the extension's upgrade actions
- The secure cookie feature would be interrupted
- With the static `rules.json` file, this could mean having to do a full extension update rather than just updating the rules separately like it is handled now.
- Redirect looping handling and other features are not necessarily addressed for replacement in declarativeNetRequest are also in question

There is already bug filed here with lengthy discussion: https://bugs.chromium.org/p/chromium/issues/detail?id=896897&desc=2

This conversation is also being brought up with other privacy oriented extensions:
https://github.com/uBlockOrigin/uBlock-issues/issues/338

**Comment by Parasrah** [comment]
Thanks for the response, glad to hear this wasn't a surprise and there have been talks internally. Would it be possible to update this issue when a decision has been reached by the team as to next steps?

**Comment by zoracon** [comment]
@Parasrah yea we will update this as a consensus is made on what to do programmatically. Right now the discussion is still young among us and other privacy extensions. [There is a working group](https://groups.google.com/a/chromium.org/forum/#!topic/chromium-extensions/veJy9uAwS00) having discussion and discourse being had with the proper channels. 

Since this is a draft the possibility for changes remain and we are hoping to extend the future of the extension's current capabilities if we clearly outline early on what issues this `declarativeNetRequest` API currently causes.

**Comment by Parasrah** [comment]
Thanks, I appreciate it, and will keep an eye on the various conversations

**Comment by pipboy96** [comment]
Related: https://github.com/EFForg/privacybadger/issues/2273.

**Comment by pipboy96** [comment]
Someone needs to pin this issue like the Privacy Badger one.

**Comment by zoracon** [comment]
Linked is the document I attached to the [Google Group](https://groups.google.com/a/chromium.org/forum/#!topic/chromium-extensions/veJy9uAwS00%5B101-125%5D) discussing Manifest V3. The general thought process is to request functionality that is missing for the declarativeNetRequest API and breaks HTTPSE as it currently stands. Will be on stand by for further responses from the group and watching out for any changes in the dNR API.



https://docs.google.com/document/d/1vzSyzrrjOJrUavfOWC5FC2PNwcBOfN2WkK_HcTdTacw/edit?usp=sharing

**Comment by pipboy96** [comment]
Related: Eloston/ungoogled-chromium#706.

**Comment by patrickschneider23** [comment]
[ `` ](url)

**Comment by pipboy96** [comment]
> [ `` ](url)

?

**Comment by zoracon** [comment]
As an update, set up general asks if declarativeNetRequest were to happen. So far some responses on what can and can't happen, but generally still watching out for _written_ updates to the potential dNR API

Public Conversation here in the Google Group.
https://groups.google.com/a/chromium.org/d/msg/chromium-extensions/veJy9uAwS00/UfJdEm2eBgAJ

**Comment by pipboy96** [comment]
@zoracon Care to delete https://github.com/EFForg/https-everywhere/issues/17268#issuecomment-473614985, https://github.com/EFForg/https-everywhere/commit/5445384b37448a796054ba555bc71f59161761ef#r32812770, https://github.com/EFForg/https-everywhere/commit/5445384b37448a796054ba555bc71f59161761ef#r32812750?

**Comment by zoracon** [comment]
> @zoracon Care to delete [#17268 (comment)](https://github.com/EFForg/https-everywhere/issues/17268#issuecomment-473614985), [5445384#r32812770](https://github.com/EFForg/https-everywhere/commit/5445384b37448a796054ba555bc71f59161761ef#r32812770), [5445384#r32812750](https://github.com/EFForg/https-everywhere/commit/5445384b37448a796054ba555bc71f59161761ef#r32812750)?

@pipboy96 Currently in the process from another issue these comments were flagged in. I like to keep topics separate in issues

**Comment by pipboy96** [comment]
@zoracon @Hainish Is there any follow-up to this?

**Comment by Hainish** [comment]
When was the last edit of this document?  Last time I checked we were waiting for a second draft to track what has changed.

**Comment by zoracon** [comment]
Right now we are still waiting for the second draft of Manifest V3. The last time the document was updated was November 18th, 2018.

Public conversation and any further questions can be posted to the public Google group.
https://groups.google.com/a/chromium.org/d/msg/chromium-extensions/veJy9uAwS00/UfJdEm2eBgAJ

But unfortunately we are stuck in the position of waiting on changes to the draft

**Comment by pipboy96** [comment]
@zoracon Any updates?

**Comment by drbcode** [comment]
Will Declarative Net Request API be able to modify request and response headers to bypass CORS and access remote content that extension has permission to access ?

We have developed and used various automation tools that heavily rely on modifying request and response headers to send automated web requests to web servers.

Therefore I am hoping that Declarative Net Request API will support 
- modifying request and response headers.
- adding new request or response headers.
- removing request or response headers.

There are countless web automation tools deployed on the web that rely on modifying request and response headers.

Apart from automation tools, there are many other legitimate use cases for modifying request or response headers, for example: chrome extension to change user agent.

At the moment there are millions of users that make use of web automation tools.

If Declarative Net Request API fails to add support for modifying headers and webRequest API is phased out then many web automation tools will stop working.

**Comment by pipboy96** [comment]
@drbcode We don't know, since we are not Chromium developers. Sorry.

**Comment by zoracon** [comment]
> @zoracon Any updates?

Looking at the docs for declarativeNetRequest, they seemed to have added some changes involving dynamic rules and redirect. The Google Group discussion so far came back with with statement:

> That said, the extensions team is currently working on a "developer preview" of the MV3 changes. The intent is to implement most of the major changes to the platform so that extension developers can start transitioning and sharing feedback about the process. We don't have an exact date for this, but we're hoping to ship it in the next few months. - Simeon - @dotproto, Extensions Developer Advocate

Right now, the best action for a HTTPS Everywhere volunteer, user, enthusiast would be to look into jumping into the discussion in the public Google Group discussing these changes. On our end, we are  discussing internally what next steps look like and doing our best to create a plan.

**Comment by pipboy96** [comment]
**UPDATE: Blocking capabilities of webRequest are still being deprecated, see** https://groups.google.com/a/chromium.org/forum/?utm_medium=email&utm_source=footer#!topic/chromium-extensions/veJy9uAwS00%5B101-125%5D and commentary by @gorhill: https://github.com/uBlockOrigin/uBlock-issues/issues/338#issuecomment-496009417.

Thanks @Giltyhub for update!

**Comment by Saiv46** [comment]
Looks like if webRequest is got removed in future, HTTPS Everywhere will come as a local proxy than an extension.

**Comment by pipboy96** [comment]
@Saiv46 I had the same idea.

**Comment by cschanaj** [comment]
Here is a Chromium blog post on the impact of the breaking change (from their perspective): https://blog.chromium.org/2019/06/web-request-and-declarative-net-request.html 

It sound like that they are going to depreciate the Web Request APIs anyawy...

**Comment by pipboy96** [comment]
@cschanaj It sounds like they are trying to justify the choice they made with arguments other than the ones that made them make the choice. They are trying to "sell" the changes to the people.

**Comment by zoracon** [comment]
Update from Mozilla:
https://blog.mozilla.org/addons/2019/09/03/mozillas-manifest-v3-faq/

**Comment by pipboy96** [comment]
The way it's formulated it seems some degree of rewriting will be necessary for extension to stay compatible even if Firefox will decide to keep `webRequestBlocking` permission.

**Comment by pipboy96** [comment]
@zoracon Can EFF (and possibly Tor Project) officially contact Mozilla, explaining them why `webRequestBlocking` should be kept? I'm sure they will listen.

**Comment by pipboy96** [comment]
Removed `chrome` and `firefox` labels since they are the only browsers we support anyway.

**Comment by zoracon** [comment]
After much back and forth between different privacy extensions and the Chrome team. They have rolled out Manifest V3 into Canary
https://groups.google.com/a/chromium.org/forum/#!msg/chromium-extensions/hG6ymUx7NoQ/TiAe0gI5AQAJ

**First steps:**

- Going to load up HTTPS Everywhere in Canary and see what breaks. (Likely everything :( )
- Next I'm probably going to move faster to clean up rulesets and run a[ fetch test](https://github.com/EFForg/https-everywhere/issues/13829) as soon as possible
- Once that is done, I am going to package trivialized rulesets for Chrome (because at this point, the more complicated rulesets will just slow the process down of migration to V3)
- Internally we are working on messaging on how we will communicate this Chrome version
- Please allow patience on other features at this time and [design implementation will work will be temporarily put on hold](https://github.com/uracreative/works/issues/29) until this is sorted out

**Unknowns**
- Still not sure how Chrome plans on allocating rulesets, only the privacy extensions mainly use this feature. But this new cap could prove to be a very big challenge
- Still not sure on the full impact of migrating from background pages to service workers. I'm very concerned on how debugging the extension will be handled now.

**Comment by pipboy96** [comment]
@zoracon Current Canary can still load Manifest V2 extension it seems.

**Comment by zoracon** [comment]
> @zoracon Current Canary can still load Manifest V2 extensions it seems.

@pipboy96 Right, they don't specify when V2 will hit EoL. To test V3, just have to change the `manifest.json` file to `"manifest_version": 3`. Of course, HTTPS Everywhere simply doesn't install in it's current state. I'll outline more tasks once I read through the migration guide.

**Comment by zoracon** [comment]
Just a quick update with where we are with this

MV3 is still highly unstable and I am working up a PoC of what the MV3 extension could look like:

Basically, it would have to run strict HTTPS only to get around the ruleset cap as much as possible.

As nice as strict HTTPS is, the volatile changes on MV3's new declarativeNetRequest API and service worker structure are very difficult and frustrating to test with at the moment. There is a bug filed here:
https://bugs.chromium.org/p/chromium/issues/detail?id=1024211

**Comment by pipboy96** [comment]
@zoracon Do you mean MV3 version will be pretty much EASE-only?

**Comment by pipboy96** [comment]
@zoracon What's the status of this right now?

**Comment by zoracon** [comment]
@pipboy96 Pretty much in the same state as when I commented here:
https://github.com/EFForg/https-everywhere/issues/17268#issuecomment-589462371

Other developments include the dNR API adding regex rulesets, ability to load different ruleset objects, and better url filtering. With the latest developments + the API finally being more available for testing, I can continue to see what a HTTPS Everywhere in MV3 looks like.

The major issue that I want to tackle is being able to load all applicable rulesets. Which still doesn't seem possible. Google still unilaterally makes changes to the API without much notice. With that said I am working with a similar extension to see what issues we come across collectively. The bigger voice we have, changes do _seem_ to happen in response to our critiques that make the API more tolerable.

However, there are still major concerns on their timeline and when exactly will MV2 will be EOL. [Which is still unknown](https://developer.chrome.com/extensions/migrating_to_manifest_v3#timeline). Hopefully we get a better sense of the timeline, but until then we are just testing changes and trying to adapt while keeping in mind to advocate for privacy extensions to survive MV3.

**Comment by zoracon** [comment]
Cross referencing my comment made here in https://github.com/EFForg/https-everywhere/issues/18069#issuecomment-694635612

>"To add to this, MV3 (#17268) is not currently prioritized. The plan is to port to MV3 once they set an actual EoL on MV2. But in order to be prepared, we have to breakdown what we currently have into more modular and compartmentalized code."
